File: cpp/380_Insert Delete GetRandom O(1).cpp
/**
 * Use a nums vector to record all the 
 **/ 
class RandomizedSet {
public:
    /** Initialize your data structure here. */
    RandomizedSet() {
        
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if (value_to_loc.find(val) == value_to_loc.end()) {
            nums.push_back(val);
            value_to_loc[val] = nums.size() - 1;
            return true;
        }
        return false;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        int size = nums.size();
        auto val_iter = value_to_loc.find(val);
        if (val_iter != value_to_loc.end()) {
            int loc = val_iter->second;
            swap(nums[size-1], nums[loc]);
            value_to_loc[nums[loc]] = loc;
            value_to_loc.erase(val);
            nums.pop_back();
            return true;
        }
        return false;        
    }
    
    /** Get a random element from the set. */
    int getRandom() {        
        return nums[uniform_int_distribution<int>(0,nums.size() - 1)(gen)];
    }
private:
    random_device rd;
    mt19937 gen{rd()};
    vector<int> nums;
    unordered_map<int, int> value_to_loc;
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * bool param_1 = obj.insert(val);
 * bool param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: java/380_Insert Delete GetRandom O(1).java
/**
 * 如果要随机，那只能是用0..n对应数字的方法。但这样删除数字的时候怎么办？简单，只要用最后一个n对应的数字覆盖掉就行了
 */

class RandomizedSet {

  private Map<Integer, Integer> idToNum = new HashMap<>();
  private Map<Integer, Integer> numToId = new HashMap<>();
  Random random = new Random();
  int n = 0;
  
  /** Initialize your data structure here. */
  public RandomizedSet() {
  }
    
  /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
  public boolean insert(int val) {
    if (!numToId.containsKey(val)) {
      idToNum.put(n, val);
      numToId.put(val, n);
      ++n;
      return true;
    } else {
      return false;
    }
  }
    
  /** Removes a value from the set. Returns true if the set contained the specified element. */
  public boolean remove(int val) {
    if (numToId.containsKey(val)) {
      int m = numToId.get(val);
      numToId.remove(val);
      idToNum.remove(m);
      if (m < n - 1) {
        idToNum.put(m, idToNum.get(n - 1));
        numToId.put(idToNum.get(n - 1), m);
        idToNum.remove(n - 1);
      }
      --n;
      return true;
    } else {
      return false;
    }
  }
    
  /** Get a random element from the set. */
  public int getRandom() {
    int randId = random.nextInt(n);
    return idToNum.get(randId);
  }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */